### DISCUSSION TITLE
(Solved)Cyclic dependency issue with ahash crate - maybe worthwhile to document.

### DISCUSSION THREAD

**Discussion by elbart** [SeaQL/sea-orm#238] [Q&A] [answered]
Dear Sea ORM creators / maintainers,

first of all, thanks a lot for all the effort you seem to spend on this nice peace of software. I was just skimming around to look for a ORM to try for a new project I am trying to build. While knowing diesel (no it's not async, so it's not really interesting to me anymore) and also trying rbatis, Sea ORM looks the most promising to be honest. I have been working with sqlx before but wished for a more high level layer for regular crud operations.

I have been trying to pull in Sea ORM in an quasi active project with openidconnect / oauth dependency, however failed to pull in Sea ORM. I got a cyclic dependency error:

```bash
> cargo build
    Updating crates.io index
error: cyclic package dependency: package `ahash v0.7.4` depends on itself. Cycle:
package `ahash v0.7.4`
    ... which is depended on by `hashbrown v0.11.2`
    ... which is depended on by `indexmap v1.7.0`
    ... which is depended on by `serde_json v1.0.67`
    ... which is depended on by `wasm-bindgen v0.2.77`
    ... which is depended on by `js-sys v0.3.54`
    ... which is depended on by `getrandom v0.2.3`
    ... which is depended on by `ahash v0.7.4
```

I was a bit puzzled and did a `cargo clean` but this did not help, too. I searched for such an Issue and found an unanswered SO post: https://stackoverflow.com/questions/68399961/how-do-i-resolve-a-cyclic-dependency-in-cargo in the same question range. 

I then wanted to do a minimal example for reproduction in a clean project ... I failed! I was unabled to reproduce, which left me puzzled. The only thing for me, left to do was to do `rm Cargo.lock && cargo clean && cargo build` and this suddenly worked (*remark: i did not change any version specs in Cargo.toml*, but I never pinned to patch versions, but usually minors). Seems like I was locked to an older version somewhere - the changes in Cargo.lock were too many to track down where the issue could have been, however it resolved the issue for me. Instead of deleting Cargo.toml I just learned, that i should have better used `cargo update` :).

Anyways, other people pulling Sea ORM in existing projects might stumble upon this issue and do not jump on the train. A simple 'cargo update' instruction for the above error somewhere in the documentation could help :)

Keep up the great work and feel free to close if too minor for you guys!

**Comment by baoyachi** [comment, marked as answer]
it's caused by this issue:[https://github.com/tkaitchuck/aHash/issues/95](https://github.com/tkaitchuck/aHash/issues/95).
Using indexmap = "~1.6.2" works as a workaround.

Sea-orm can not solved this question.

**Comment by elbart** [comment]
@baoyachi thanks for the quick reply! I never claimed that sea orm can solve the issue, sorry if that somehow sounded like that (which is why I put solved in the ticket title). I just wanted to point out that this issue happened to me while adding sea orm to the project and never happend before with any other crate I added. Therefore I thought it might be good to document this here, too, just if someone else stumbles upon this same issue. And if you guys want, you could add a short remark to your already great documentation. No must have, it just came to my mind.

**Comment by tyt2y3** [comment]
Thank you @baoyachi for the help. 

So according to the issue it seems that the problem manifests when the serde feature on wasm-bindgen is enabled. Did you actually have wasm in your dependency tree? Or is it another dependency combo which caused this?

I have never encountered it indeed. It might be some case that the Rust library team is not aware of, and so might be worthwhile to boil down the offending crate combo.

**Comment by omofolarin** [comment]
cyclic package dependency: package `serde_json v1.0.78` depends on itself. Cycle:
package `serde_json v1.0.78`
    ... which satisfies dependency `serde_json = "^1.0"` of package `wasm-bindgen v0.2.78`
    ... which satisfies dependency `wasm-bindgen = "^0.2.78"` of package `js-sys v0.3.55`
    ... which satisfies dependency `js-sys = "^0.3"` of package `getrandom v0.2.3`
    ... which satisfies dependency `getrandom = "^0.2.3"` of package `ahash v0.7.6`
    ... which satisfies dependency `ahash = "^0.7.0"` of package `hashbrown v0.11.2`
    ... which satisfies dependency `hashbrown = "^0.11"` of package `indexmap v1.8.0`
    ... which satisfies dependency `indexmap = "^1.5"` of package `serde_json v1.0.78`
    ... which satisfies dependency `serde_json = "^1"` of package `stdweb-internal-macros v0.2.9`
    ... which satisfies dependency `stdweb-internal-macros = "=0.2.9"` of package `stdweb v0.4.20`
    ... which satisfies dependency `stdweb = "^0.4.20"` of package `time v0.2.27`
    ... which satisfies dependency `time = "^0.2.11"` of package `cookie v0.14.4`
    ... which satisfies dependency `cookie = "^0.14.1"` of package `actix-http v2.2.1`
    ... which satisfies dependency `actix-http = "^2.2.0"` of package `actix-web v3.3.3`
    ... which satisfies dependency `actix-web = "^3.0.0"` of package `actix-cors v0.5.4`

The issue persists

**Reply by billy1624** [reply]
Hey @omofolarin, problem solved?

**Comment by omofolarin** [comment]
Not yet still trying to resolve it. I posted it on a rust channel, someone said he did a cargo check on my cargo.toml and didn't face the issue. So I'm my rust version to the latest release. I'll report back after thatOn Jan 25, 2022 5:43 AM, Billy Chan ***@***.***> wrote:
Hey @omofolarin, problem solved?

**Reply by billy1624** [reply]
Can you run `cargo update` then compile it again with `cargo build`?

**Reply by omofolarin** [reply]
@billy1624 the issue seems to be related to cargo workspace. I excluded the project from the workspace and the build was successful.

**Reply by billy1624** [reply]
Can you share a bit more on how you organize the workspace?

**Reply by omofolarin** [reply]
<img width="1018" alt="Screenshot 2022-01-26 at 00 33 17" src="https://user-images.githubusercontent.com/7985548/151078086-62fa23a5-0267-4fd4-bad3-fe3e6c104950.png">

updating  oauth2 on another project in the same cargo workspace fixed the issue. And there was no longer a need to pin (at least for now)

**Reply by billy1624** [reply]
Thanks for the update @omofolarin. Feel free to ping us here or in Discord :) Have fun with Rust!
